[Draft] Introduce `torch.distributed.nn.utils.init_meta()`

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66317

This PR introduces the `init_meta()` function that enables constructing `torch.nn.Module` instances on the meta device for inspection purposes. It is meant to be used if the size of a module is too big to fit on a single machine and you want to inspect it without instantiating.

Internally ``init_meta()`` forces all parameters, buffers, and other tensors within the scope of the module and its sub-modules to use the meta device regardless of the actual device type. The returned module can be used for inspection purposes and afterwards its ``materialize()`` method can be called to construct a fully-instantiated module.

```
>>> import torch.distributed.nn
>>> m = torch.distributed.nn.utils.init_meta(torch.nn.Linear, 5, 1)
>>> m.weight
Parameter containing:
tensor(..., device='meta', requires_grad=True)
>>> m = m.materialize()
>>> m.weight
Parameter containing:
tensor([[-1.4677e+24,  4.5915e-41,  1.4013e-45,  0.0000e+00,
             -1.4677e+24, 4.5915e-41]], requires_grad=True)
```

However note that ``init_meta()`` uses a "best effort" algorithm and is not guaranteed to succeed if the underlying module's implementation cannot be mapped to the meta device. If you are the module author, you can use the `is_meta_init()` function to find out whether your module is being instantiated in the scope of a ``init_meta()`` call and use an alternate logic if required.

```
class MyModule(Module):
    def __init__(self):
        if torch.distributed.nn.utils.is_meta_init():
            self.myparam = torch.empty([10,10], device="meta")
        else:
            self.myparam = load_myparam()
```



**init_meta() vs. skip_init()**

Although they sound similar `init_meta()` and `skip_init()` serve different purposes. `init_meta()` is meant for lazy model materialization and targets model users, while ``skip_init()`` is meant for optimized module initialization and targets model authors.

Technical differences between the two functions are:

- `init_meta()` returns a module where all its parameters and buffers reside on the meta device, while `skip_init()` returns a module where some or all its parameters or buffers are empty tensors allocated on a real device.
- `init_meta()` does not require any code changes in the module, while `skip_init()` expects the module and its submodules to take a `device` argument. This means ``skip_init()`` is intrusive, but also guaranteed to work; on the other hand ``init_meta()`` can potentially fail to construct a module if the module initializes its state in an unconventional way that conflicts with the meta device.

Differential Revision: [D31504090](https://our.internmc.facebook.com/intern/diff/D31504090/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

cbalioglu committed:
Introduce `torch.distributed.nn.utils.describe()`

This PR introduces the `describe()` function that enables constructing `torch.nn.Module` instances on the meta device for inspection purposes. It is meant to be used if the size of a module is too big to fit on a single machine and you want to inspect it without instantiating.

Internally ``describe()`` forces all parameters, buffers, and other tensors within the scope of the module and its sub-modules to use the meta device regardless of the actual device type. The returned module can be used for inspection purposes and afterwards its ``materialize()`` method can be called to construct a fully-instantiated module.

```
>>> import torch.distributed.nn
>>> m = torch.distributed.nn.utils.describe(torch.nn.Linear, 5, 1)
>>> m.weight
Parameter containing:
tensor(..., device='meta', requires_grad=True)
>>> m = m.materialize()
>>> m.weight
Parameter containing:
tensor([[-1.4677e+24,  4.5915e-41,  1.4013e-45,  0.0000e+00,
             -1.4677e+24, 4.5915e-41]], requires_grad=True)
```

However note that ``describe()`` uses a "best effort" algorithm and is not guaranteed to succeed if the underlying module's implementation cannot be mapped to the meta device. If you are the module author, you can use the `is_described()` function to find out whether your module is being instantiated in the scope of a ``describe()`` call and use an alternate logic if required.

```
class MyModule(Module):
    def __init__(self):
        if torch.distributed.nn.utils.is_described():
            self.myparam = torch.empty([10,10], device="meta")
        else:
            self.myparam = load_myparam()
```



**describe() vs. skip_init()**

Although they sound similar `describe()` and `torch.nn.utils.skip_init()` serve different purposes. `describe()` is meant for lazy model materialization and targets model users, while ``skip_init()`` is meant for optimized module initialization and targets model authors.

Technical differences between the two functions are:

- `describe()` returns a module where all its parameters and buffers reside on the meta device, while `skip_init()` returns a module where some or all its parameters or buffers are empty tensors allocated on a real device.
- `describe()` does not require any code changes in the module, while `skip_init()` expects the module and its submodules to take a `device` argument. This means ``skip_init()`` is intrusive, but also guaranteed to work; on the other hand ``describe()`` can potentially fail to construct a module if the module initializes its state in an unconventional way that conflicts with the meta device.

Differential Revision: [D31504090](https://our.internmc.facebook.com/intern/diff/D31504090/)

[ghstack-poisoned]

cbalioglu commented:
> `describe` seems a very generic name :( also, if it's deemed more useful, maybe better put it in a more generic `torch.nn.utils` namespace next to `skip_init`?

Thanks for the feedback @vadimkantorov! There are ongoing discussions about the right place to put this function. I assume next week both its final name and final location will become clear.

cbalioglu committed:
Update on "[Draft] Introduce `torch.distributed.nn.utils.describe()`"


This PR introduces the `describe()` function that enables constructing `torch.nn.Module` instances on the meta device for inspection purposes. It is meant to be used if the size of a module is too big to fit on a single machine and you want to inspect it without instantiating.

Internally ``describe()`` forces all parameters, buffers, and other tensors within the scope of the module and its sub-modules to use the meta device regardless of the actual device type. The returned module can be used for inspection purposes and afterwards its ``materialize()`` method can be called to construct a fully-instantiated module.

```
>>> import torch.distributed.nn
>>> m = torch.distributed.nn.utils.describe(torch.nn.Linear, 5, 1)
>>> m.weight
Parameter containing:
tensor(..., device='meta', requires_grad=True)
>>> m = m.materialize()
>>> m.weight
Parameter containing:
tensor([[-1.4677e+24,  4.5915e-41,  1.4013e-45,  0.0000e+00,
             -1.4677e+24, 4.5915e-41]], requires_grad=True)
```

However note that ``describe()`` uses a "best effort" algorithm and is not guaranteed to succeed if the underlying module's implementation cannot be mapped to the meta device. If you are the module author, you can use the `is_described()` function to find out whether your module is being instantiated in the scope of a ``describe()`` call and use an alternate logic if required.

```
class MyModule(Module):
    def __init__(self):
        if torch.distributed.nn.utils.is_described():
            self.myparam = torch.empty([10,10], device="meta")
        else:
            self.myparam = load_myparam()
```



**describe() vs. skip_init()**

Although they sound similar `describe()` and `torch.nn.utils.skip_init()` serve different purposes. `describe()` is meant for lazy model materialization and targets model users, while ``skip_init()`` is meant for optimized module initialization and targets model authors.

Technical differences between the two functions are:

- `describe()` returns a module where all its parameters and buffers reside on the meta device, while `skip_init()` returns a module where some or all its parameters or buffers are empty tensors allocated on a real device.
- `describe()` does not require any code changes in the module, while `skip_init()` expects the module and its submodules to take a `device` argument. This means ``skip_init()`` is intrusive, but also guaranteed to work; on the other hand ``describe()`` can potentially fail to construct a module if the module initializes its state in an unconventional way that conflicts with the meta device.

Differential Revision: [D31504090](https://our.internmc.facebook.com/intern/diff/D31504090/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

anj-s commented:
Thank you for this feature! This is very much needed for initializing large models.

cbalioglu committed:
Update on "[Draft] Introduce `torch.distributed.nn.utils.init_meta()`"


This PR introduces the `init_meta()` function that enables constructing `torch.nn.Module` instances on the meta device for inspection purposes. It is meant to be used if the size of a module is too big to fit on a single machine and you want to inspect it without instantiating.

Internally ``init_meta()`` forces all parameters, buffers, and other tensors within the scope of the module and its sub-modules to use the meta device regardless of the actual device type. The returned module can be used for inspection purposes and afterwards its ``materialize()`` method can be called to construct a fully-instantiated module.

```
>>> import torch.distributed.nn
>>> m = torch.distributed.nn.utils.init_meta(torch.nn.Linear, 5, 1)
>>> m.weight
Parameter containing:
tensor(..., device='meta', requires_grad=True)
>>> m = m.materialize()
>>> m.weight
Parameter containing:
tensor([[-1.4677e+24,  4.5915e-41,  1.4013e-45,  0.0000e+00,
             -1.4677e+24, 4.5915e-41]], requires_grad=True)
```

However note that ``init_meta()`` uses a "best effort" algorithm and is not guaranteed to succeed if the underlying module's implementation cannot be mapped to the meta device. If you are the module author, you can use the `is_meta_init()` function to find out whether your module is being instantiated in the scope of a ``init_meta()`` call and use an alternate logic if required.

```
class MyModule(Module):
    def __init__(self):
        if torch.distributed.nn.utils.is_meta_init():
            self.myparam = torch.empty([10,10], device="meta")
        else:
            self.myparam = load_myparam()
```



**init_meta() vs. skip_init()**

Although they sound similar `init_meta()` and `skip_init()` serve different purposes. `init_meta()` is meant for lazy model materialization and targets model users, while ``skip_init()`` is meant for optimized module initialization and targets model authors.

Technical differences between the two functions are:

- `init_meta()` returns a module where all its parameters and buffers reside on the meta device, while `skip_init()` returns a module where some or all its parameters or buffers are empty tensors allocated on a real device.
- `init_meta()` does not require any code changes in the module, while `skip_init()` expects the module and its submodules to take a `device` argument. This means ``skip_init()`` is intrusive, but also guaranteed to work; on the other hand ``init_meta()`` can potentially fail to construct a module if the module initializes its state in an unconventional way that conflicts with the meta device.

Differential Revision: [D31504090](https://our.internmc.facebook.com/intern/diff/D31504090/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

dzhulgakov reviewed (commented):
I still feel it's best to put it in just `torch.nn.utils` as it's not really distributed-specific

tchaton commented:
Hey @cbalioglu,

Find there is an issue with deepcopy and init_meta. Here is a reproducible script.

```py
import torch
from copy import deepcopy

class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        # this works
        default = torch.nn.parameter.Parameter(torch.tensor(1.0))
        deepcopy(default)

        # this doesn't
        default = torch.tensor(1.0)
        deepcopy(default)

torch.distributed.nn.utils.init_meta(TestModule)

"""
  deepcopy(default)
  File "/Users/thomas/.pyenv/versions/3.8.5/lib/python3.8/copy.py", line 153, in deepcopy
    y = copier(memo)
  File "/Users/thomas/Documents/GitHub/pytorch-lightning/.venv/lib/python3.8/site-packages/torch/_tensor.py", line 120, in __deepcopy__
    new_tensor.set_(new_storage, self.storage_offset(), self.size(), self.stride())
RuntimeError: Missing cases in 'toPyObject'! Can't convert Storage to a Python object
"""
```

cbalioglu commented:
Hey @tchaton, this looks very similar to #66444 that I opened a short while ago. It seems to be an issue related to the new Python mode API. @zou3519, any update on that?

cbalioglu commented:
Related PRs that I have merged recently to unblock this work: #67055, #67032, #66486. WIP: #65330

cbalioglu commented:
Thanks to @stas00 for https://github.com/huggingface/transformers/pull/14447/files which removes another obstacle in large model materialization work.
